compiler/src/dotty/tools/dotc/transform/Erasure.scala

```diff
@@ -1044,7 +1045,17 @@ object Erasure {
 
     override def typedClassDef(cdef: untpd.TypeDef, cls: ClassSymbol)(using Context): Tree =
       if cls.is(Flags.Erased) then erasedDef(cls)
-      else super.typedClassDef(cdef, cls)
+      else
+        val typedTree@TypeDef(name, impl @ Template(constr, _, self, _)) = super.typedClassDef(cdef, cls): @unchecked
+        // In the case where a trait extends a class, we need to strip any non trait class from the signature
+        // and accept the first one (see tests/run/mixins.scala)
+        val newParents = impl.parents.head :: impl.parents.tail.filterConserve: tree =>
+          def isTraitConstructor = tree match
+            case Trees.Block(_, expr) => // Specific management for trait constructors (See #9216)
+              expr.symbol.isConstructor && expr.symbol.owner.is(Flags.Trait)
+            case _ => tree.symbol.isConstructor && tree.symbol.owner.is(Flags.Trait)
+          tree.symbol.is(Flags.Trait) || isTraitConstructor
+        cpy.TypeDef(typedTree)(rhs = cpy.Template(impl)(parents = newParents))
```

Comment:
Currently `Template` and `TypeDef` will always create a new instance of the tree. We can avoid that if we do the following. Note that not changing the parents will be the most common case.
```diff
- val newParents = impl.parents.head :: impl.parents.tail.filterConserve: tree =>
+ val newTraits = impl.parents.tail.filterConserve: tree =>
    ...
+ val newParents = 
+  if impl.parents.tail eq newTraits then impl.parents 
+  else impl.parents.head :: newTraits
  cpy.TypeDef(typedTree)(rhs = cpy.Template(impl)(parents = newParents)
```